File: app/src/androidTest/java/com/duchuyctlk/helpers/matchers/CustomMatchers.java
package com.duchuyctlk.helpers.matchers;

import android.view.View;
import android.widget.Spinner;

import com.duchuyctlk.Constant;
import com.duchuyctlk.widget.PopupDismissCatchableSpinner;

import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;

/**
 * Created by huy.duc on Apr 8th 2016
 */
public class CustomMatchers {

    public static Matcher<View> withSpinnerMode(final int expectedMode) {
        return new TypeSafeMatcher<View>() {

            @Override
            public boolean matchesSafely(View view) {
                if (!(view instanceof PopupDismissCatchableSpinner)) {
                    return false;
                }

                PopupDismissCatchableSpinner spinner = (PopupDismissCatchableSpinner) view;
                return spinner.getSpinnerMode() == expectedMode;
            }

            @Override
            public void describeTo(Description description) {
                String strExpectedMode = expectedMode == Spinner.MODE_DIALOG ? Constant.STR_MODE_DIALOG :
                        (expectedMode == Spinner.MODE_DROPDOWN ? Constant.STR_MODE_DROPDOWN :
                                String.valueOf(-1));
                description.appendText("with spinner mode: " + strExpectedMode);
            }
        };
    }
}

File: app/src/androidTest/java/com/duchuyctlk/helpers/matchers/CustomMatchers.kt
package com.duchuyctlk.helpers.matchers

import android.view.View
import android.widget.Spinner.MODE_DIALOG
import android.widget.Spinner.MODE_DROPDOWN
import com.duchuyctlk.Constant.STR_MODE_DIALOG
import com.duchuyctlk.Constant.STR_MODE_DROPDOWN
import com.duchuyctlk.widget.PopupDismissCatchableSpinner
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.TypeSafeMatcher

object CustomMatchers {
    fun withSpinnerMode(expectedMode: Int): Matcher<View> {
        return object : TypeSafeMatcher<View>() {
            override fun matchesSafely(view: View?): Boolean {
                if (view !is PopupDismissCatchableSpinner) {
                    return false
                }
                return view.spinnerMode == expectedMode
            }

            override fun describeTo(description: Description) {
                val strExpectedMode = when (expectedMode) {
                    MODE_DIALOG -> STR_MODE_DIALOG
                    MODE_DROPDOWN -> STR_MODE_DROPDOWN
                    else -> (-1).toString()
                }
                description.appendText("with spinner mode: $strExpectedMode")
            }
        }
    }
}
